teps/0111-replace-volumes-with-workspaces.md

```diff
@@ -0,0 +1,368 @@
+- Support for [CSI volumes](#support-csi-volumes-in-workspace-bindings).
+There's a clear use case for this feature ([#4446](https://github.com/tektoncd/pipeline/issues/4446)).
+
+Because existing Workspace binding options support most CI/CD use cases, adding more types of volumes shouldn't be a blocker for replacing
+volumes with Workspaces.
+
+#### Support all types of Volumes in WorkspaceBinding
+
+In this solution, we would allow workspace bindings to specify any kind of volumeSource:
+
+```go
+import corev1 "k8s.io/api/core/v1"
+
+type WorkspaceBinding struct {
+	// Name is the name of the workspace populated by the volume.
+	Name string `json:"name"`
+	// SubPath is an optional directory on the volume which should be used for this binding
+	SubPath string `json:"subPath,omitempty"`
+  // VolumeClaimTemplate is a template for a claim that will be created in the same namespace.
+	// The PipelineRun controller is responsible for creating a unique claim for each instance of PipelineRun.
+	// +optional
+	VolumeClaimTemplate *corev1.PersistentVolumeClaim `json:"volumeClaimTemplate,omitempty"`
+  // VolumeSource represents the location and type of the mounted volume.
+	corev1.VolumeSource `json:",inline" protobuf:"bytes,2,opt,name=volumeSource"`
+}
+```
+
+Users could not specify more than one VolumeSource, or a VolumeClaimTemplate and a VolumesSource.
+This would break Go libraries compatibility but not the API.
+
+Pros:
+- Users do not lose the ability to mount any kinds of volumes to their `TaskRun` pods.
+
+Cons:
+- There's no clear use case for many of these types of Volumes in Tekton.
+
+#### Support hostPath volumes in Workspace Bindings
+
+This solution augments the proposed solution (removing volume-related fields from the API)
+with new Workspace Bindings to support advanced Docker-in-Docker use cases.
+
+In this solution, a new field be added to [WorkspaceBinding](https://github.com/tektoncd/pipeline/blob/342ed5237f3bd3273485672426472194a186c02f/pkg/apis/pipeline/v1beta1/workspace_types.go#L54)
+to support hostPath volumes:
+
+```go
+import corev1 "k8s.io/api/core/v1"
+
+type WorkspaceBinding struct {
+	// Name is the name of the workspace populated by the volume.
+	Name string `json:"name"`
+	// SubPath is an optional directory on the volume which should be used for this binding
+	SubPath string `json:"subPath,omitempty"`
+  
+  // ... existing volume sources such as emptyDir, secret, PVC
+
+  // HostPath represents a pre-existing file or directory on the host
+	// machine that is directly exposed to the container.
+  HostPath *corev1.HostPathVolumeSource
+}
+```
+
+Pros:
+- Supports more advanced docker-in-docker use cases.
+
+Cons:
+- Not necessary for most CI/CD use cases.
+- Tasks that require hostPath would likely want to declare this dependency at authoring time rather than at runtime.
```

Comment:
Looking at both the internal and this hostPath use case, it seems like there are aa few use cases where the task authors need to provide the volume/workspace type they need and workspaces today only allow taskrun authors to supply these

Reply:
agreed! I think this is covered under the "task-internal" workspaces option